### PR TITLE
Create Application: Snap To Next Clicked Surface

### PR DESCRIPTION
**Pull request by AleziaKurdis**
This PR adds 4 new actions to the Create Applications:

image

**"Rotate As the Next Clicked Surface"**

When this action is performed on an entity or a selection of entities, then it will be rotated equally as the next surface you are going to click. (Based on the Surface Normal)
This functionality is mainly useful to put anything on a wall. Without this, it was difficult to do so with enough accuracy.
Here's a demo:
https://youtu.be/V8Y8uxQ39eA

**"Rotate 90° on X axis" / "Rotate 90° on Y axis" / "Rotate 90° on Z axis"**

As a complement of the "**Rotate As the Next Clicked Surface**" action, these 3 quick rotate actions as been added, to make it easier to rotate an entity without losing its original rotation precision. (Hotkeys has been also given to these actions as doing 90-degree rotation is frequent and more direct)

Please test this with care in "Desktop" and "HMD" modes independently. _(The code is different between these two.)_

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720.tar_ec55884d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720_ec55884d.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720.tar_ec55884d.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720_ec55884d.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720_ec55884d.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e356720_ec55884d.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-e356720_ec55884d.apk

File: scripts/system/create/edit.js [hunk 1]
```diff
@@ -1106,25 +1108,41 @@ function findClickedEntity(event) {
...
     var result;
-
-    if (iconResult.intersects) {
-        result = iconResult;
-    } else if (entityResult.intersects) {
-        result = entityResult;
-    } else {
+    if (expectingRotateAsClickedSurface) {
+        if (!SelectionManager.hasSelection() || !SelectionManager.hasUnlockedSelection()) {
+                audioFeedback.rejection();
```
Comment (ctrlaltdavid) [inline scripts/system/create/edit.js:1113]: Indenting in this if-else is off.

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js [hunk 2]
```diff
@@ -103,10 +103,25 @@ SelectionManager = (function() {
                 if (wantDebug) {
                     print("setting selection to " + messageParsed.entityID);
                 }
-                if (hmdMultiSelectMode) {
-                    that.addEntity(messageParsed.entityID, true, that);
+                if (expectingRotateAsClickedSurface) {
+                    if (!SelectionManager.hasSelection() || !SelectionManager.hasUnlockedSelection()) {
+                            audioFeedback.rejection();
```
Comment (ctrlaltdavid) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:108]: Indenting in this if-else is off.

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js [hunk 3]
```diff
@@ -2377,7 +2392,48 @@ SelectionDisplay = (function() {
...
+        //axis is a string and expect "X", "Y" or "Z"
+        if (!SelectionManager.hasSelection() || !SelectionManager.hasUnlockedSelection()) {
+            audioFeedback.rejection();
+            Window.notifyEditError("You have nothing selected, or the selection is locked.");
+        } else {
+            var currentRotation, axisRotation;
+            SelectionManager.saveProperties();
+            if (selectionManager.selections.length === 1 && spaceMode === SPACE_LOCAL) {
+                currentRotation = SelectionManager.localRotation;
+            }else{
```
Comment (ctrlaltdavid) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:2414]:
```suggestion
            } else {
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127_0db2d6f0.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127.tar_0db2d6f0.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127.tar_0db2d6f0.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127_0db2d6f0.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127_0db2d6f0.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-e5e1127_0db2d6f0.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-e5e1127_0db2d6f0.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3_ff7cf2fc.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3.tar_ff7cf2fc.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3_ff7cf2fc.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3.tar_ff7cf2fc.Z

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3_ff7cf2fc.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-0f516e3_ff7cf2fc.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-0f516e3_ff7cf2fc.exe

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by Aitolda** [comment]
Confirmed working in Desktop, but my Quest is misbehaving at the moment.  Can someone else give it a go in VR?

**Comment by SilverfishVR** [comment]
align and rotate works in desktop and VR equally

* in VR, the _Edit_ dropdown is now about half a line too long to fit in the tablet, not sure what to do about that since I don't want to loose any of the functions in there, slightly higher resolution (dpi) on the would do the trick and be desirable overall but that seems out of scope for this 🤔 



* aligning the object with +Z facing away from the clicked surface is a bit counterintuitive to me, I get that it is probably correct in relation to the clicked surface normal but if I load something like this [Teapot](https://silverfish-freestuff.s3.eu-north-1.amazonaws.com/Misc/TeaSet/Pot/Teapot_hifi.fbx) it comes in right way up (Y up) but aligning it to the ground flips it on it's side, then again, if it was a picture I wanted to place on a wall it works 🤷‍♂️ not a deal breaker.

* odd behavior when aligning to completely horizontal surfaces: create a cube and a cone, don't rotate them. Select the cone and align it to the cubes top surface. it is rotated 90° on X axis as expected, but also 135° on the Y axis (45° if you click the bottom face of the cube). if you rotate the cube even a tiny bit it aligns at nice 90° angles

All that said, the functions are most useful for matching up odd angles, and for that, it works great so definitely a quality-of-life improvement 👍

**Comment by AleziaKurdis** [comment]
**The Menu**
The menu will definitely have to be reworked as we will need room for additional items soon.
I was thinking to have submenus for "Edit > Move" and "Edit > Rotation", to contain related actions (Ideally actions with hotkeys since these will be a level deeper)
I'm considering also to bring back the red "Delete" button, after many months, I think we can agree that it's a very used functionality, enough to be in top layer, and everyone understands the trash icon without need any tooltip.
**I added this to my roadmap.** _(I think we can live as it is for now. For at least one or 2 release, we have the hotkeys, and if I remember well, it's still possible to trigger the last item of the menu, even if it's not a big deal visually)_


**+Z**
This was intentional. One of the major use for this functionality is to place "image" and "web" entities on a wall. Sadly, their right side is facing +Z instead of -Z. I had to chose what would be the most useful (or the less absurd to see) between have the entity with the front in the normal direction or have a feature that puts images or web pages inverted on the surface. 
(Note that the three 90degree rotation tools have been added to address that kind of problem where the model is not on the axis that is expected. We could not prevent all the use cases anyway.)
If you have a different view about this let me know. 

**World mode and Multiple Selection**
The feature might do weird things in those 2 situations. (actually, in World mode since Multiple Selection force that mode)
In this mode we lost the reference base on the model. (I even consider at some point to disable the feature in those conditions)
There is no miracle to do here I think (At least not a simple one).

**Adding the position to this feature**
I have recently explored the possibility to have the entity's position set at the same time. To have the entity moved to the clicked surface in addition to have the rotation fitting the surface normal. It was working fine for a single selection, but support this with the multiple selection it would need a redesign (and it's not very simple), because it needs to be addressed at the same time as the rotation computation. This is why I have not insisted to finalize the name for the action now.
**But it might be a good idea to finalize the current name** since I'm not sure that it worth the investment for now. (Personnaly, I don't know for you, I did have hard time to adjust the rotation, but not really to match the position. And when the entity are too closed from the surface, even with a perfect rotation, if you take some distance, a part of the image disappears in the wall, even with a 1 cm of tolerance.)
I would plan to set the right name when I will adjust the menu.

**Comment by Aitolda** [comment]
> I'm considering also to bring back the red "Delete" button, after many months, I think we can agree that it's a very used functionality

I fully support this :).  I miss it.

**Comment by HifiExperiments** [comment]
does this fully address https://github.com/vircadia/vircadia/issues/322 ?

**Comment by AleziaKurdis** [comment]
> does this fully address #322 ?

Not yet.
I worked on a "**Snap to next Clicked Surface**" that would do the re-position in addition to adjust the Rotation based on the normal.
But I have not pushed that fix yet since I have inconsistencies with "**World Mode**" and "**Multiple Selection**".

The more I think about this, the more I consider to have this functionality available only for "**Local Mode / Single Selection**".
Because the use case for "**World Mode**" and "**Multiple Selection**" is less there:
- In "**World Mode/Single Selection**", there is not too much decent alignment to have based on the bounding box. 
- "**Multiple Selection**" is always in "**World Mode**", so the same kind of thing happens here.

The use case that is really helpful would be a "**Single Selection**" in "**Local Mode**". (which can be a parent to not)

If I have an agreement on this, I would see to push the fix after have blocked its usage in "**World Mode**".

**Comment by AleziaKurdis** [comment]
Please, Put this on **Hold**

**Comment by AleziaKurdis** [comment]
This last modification changes "**Rotate As Next Clicked Surface**" for "**Snap To Next Clicked Surface**"

image

**Instead of only setting the rotation as the clicked surface normal,
it is now also moving the entities to the surface it-self.
Please, test it. Use different mode (Local/World), single and multiple selections.**



**FUNCTIONAL DECISION STILL TO BE MADE**:
This functionality should works correctly in "**Local Mode**". But with the "**World Mode**" (which includes the multiple selections), if you do the action on a selection that is not orthogonal, you can get weird behavior since this is the bounding box (in world reference) that is used to snap to the surface. 
**From there we have 2 options:**
- We can block this functionality for the World Mode
- We can keep it as it is if there is more benefit to keep it for the part of the World Mode that can work.

_Please, use the feature and see if there is a plus value to keep the behavior for the World mode/multiple selections. We could decide based on that evaluation.
(Let me know if you need more detail about this. )_


**KNOWN ISSUES**:
Edit menu is too full for the VR UI, but still functional. 
_This will needs some redesign **that won't be part of this PR**. The Entity List toolbar would need to be reorganized and the menu would need to support submenus._

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-f793d20.tar_a9c2a3d4.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-f793d20_a9c2a3d4.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-f793d20.tar_a9c2a3d4.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-f793d20_a9c2a3d4.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-f793d20_a9c2a3d4.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-f793d20_a9c2a3d4.apk

**Comment by Aitolda** [comment]
This needs to be rebuilt for windows.

**Comment by AleziaKurdis** [comment]
I reorganized the header of the entity list with a new menubar
I replaced the radius filter label by an icon (radar) and use the saved space for the new menubar.
I bring back the red delete button (it is useful enough to be in the toolbar and don't need any text to explain what it does.)
We have now lot of room to add new action (in menu and in toolbar) for future implementations.
image

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390.tar_70781fea.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390.tar_70781fea.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390_70781fea.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390_70781fea.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-5f20390_70781fea.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390_70781fea.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-5f20390_70781fea.exe

**Comment by Aitolda** [comment]
Nice work! Everything seems to be working as intended.

File: scripts/system/html/css/edit-style.css [hunk 4]
```diff
@@ -1340,35 +1346,59 @@ div#grid-section, body#entity-list-body {
...
     float: right;
     margin-right: -168px;
-    top: -45px;
+    top: -28px;
 }
-#filter-radius-and-unit label {
-    margin-left: 2px;
+#entity-list-menubar {
+    margin: 0px -8px 6px -8px;
+    padding: 0px 8px 0px 8px;;
```
Comment (digisomni) [inline scripts/system/html/css/edit-style.css:1367]:
```suggestion
    padding: 0px 8px 0px 8px;
```

File: scripts/system/controllers/controllerModules/inEditMode.js [hunk 5]
```diff
@@ -80,7 +83,9 @@ Script.include("/~/system/libraries/utils.js");
                             Messages.sendLocalMessage(this.ENTITY_TOOL_UPDATES_CHANNEL, JSON.stringify({
                                 method: "selectEntity",
                                 entityID: this.selectedTarget.objectID,
-                                hand: hand
+                                hand: hand,
+                                surfaceNormal: this.selectedTarget.surfaceNormal,
+                                intersection: this.selectedTarget.intersection,
```
Comment (digisomni) [inline scripts/system/controllers/controllerModules/inEditMode.js:88]:
```suggestion
                                intersection: this.selectedTarget.intersection
```

File: scripts/system/create/audioFeedback/audioFeedback.js [hunk 6]
```diff
@@ -30,5 +31,12 @@ audioFeedback = (function() {
         });
     }
 
+    that.action = function() { //Play an action sound
+        var injector = Audio.playSound(actionSound, {
+             "volume": 0.3,
+            "localOnly": true
```
Comment (digisomni) [inline scripts/system/create/audioFeedback/audioFeedback.js:37]:
```suggestion
             "localOnly": true
```

**Review by digisomni — APPROVED** [review]
CR: looks good to me. :) Can we get +1 tester on this?

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916_f6aaa000.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916.tar_f6aaa000.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916.tar_f6aaa000.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916_f6aaa000.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916_f6aaa000.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Quest-Alpha-PR932-129c916_f6aaa000.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR932/Vircadia-Alpha-PR932-129c916_f6aaa000.exe

**Comment by SilverfishVR** [comment]
seems to work very well indeed, very useful addition, I approve 👍

**Review by digisomni — APPROVED** [review]

**Pull request merged**